parquet-cli/src/test/java/org/apache/parquet/cli/commands/TransCompressionCommandTest.java

```diff
@@ -0,0 +1,271 @@
+    convertRecordByRecord(CompressionCodecName.valueOf("ZSTD"), new Path(inputFile), new Path(outputFile));
+    long durationRecord = System.currentTimeMillis() - start;
+
+    // The TransCompressionCommand is ~5 times faster than translating record by record
+    Assert.assertTrue(durationTrans < durationRecord);
+  }
+
+  private void testInternal(String srcCodec, String destCodec, ParquetProperties.WriterVersion writerVersion, int pageSize) throws Exception {
+    int numRecord = 1000;
+    String inputFile = createParquetFile("input", srcCodec, numRecord, writerVersion, pageSize);
+    String outputFile = createTempFile("output_trans");
+
+    TransCompressionCommand command = new TransCompressionCommand(createLogger());
+    command.setConf(new Configuration());
+    command.input = inputFile;
+    command.output = outputFile;
+    command.codec = destCodec;
+    command.run();
+
+    validateColumns(inputFile, numRecord);
+    validMeta(inputFile, outputFile);
+    validColumnIndex(inputFile, outputFile);
+  }
+
+  private void convertRecordByRecord(CompressionCodecName codecName, Path inpath, Path outpath) throws Exception {
+    ParquetMetadata metaData = ParquetFileReader.readFooter(conf, inpath, NO_FILTER);
+    MessageType schema = metaData.getFileMetaData().getSchema();
+    HadoopInputFile inputFile = HadoopInputFile.fromPath(inpath, conf);
+    ParquetReadOptions readOptions = HadoopReadOptions.builder(conf).build();
+
+    conf.set(GroupWriteSupport.PARQUET_EXAMPLE_SCHEMA, schema.toString());
+    ExampleParquetWriter.Builder builder = ExampleParquetWriter.builder(outpath).withConf(conf).withCompressionCodec(codecName);
+
+    ParquetWriter parquetWriter = builder.build();
+
+    PageReadStore pages;
+    ParquetFileReader reader = new ParquetFileReader(inputFile, readOptions);
+
+    while ((pages = reader.readNextRowGroup()) != null) {
+      long rows = pages.getRowCount();
+      MessageColumnIO columnIO = new ColumnIOFactory().getColumnIO(schema);
+      RecordReader recordReader = columnIO.getRecordReader(pages, new GroupRecordConverter(schema));
+
+      for (int i = 0; i < rows; i++) {
+        SimpleGroup simpleGroup = (SimpleGroup) recordReader.read();
+        parquetWriter.write(simpleGroup);
+      }
+    }
+
+    parquetWriter.close();
+  }
+
+  private void validateColumns(String inputFile, int numRecord) throws IOException {
+    ParquetReader<Group> reader = ParquetReader.builder(new GroupReadSupport(), new Path(inputFile)).withConf(conf).build();
+    for (int i = 0; i < numRecord; i++) {
+      Group group = reader.read();
+      assertTrue(group.getLong("DocId", 0) < 1000);
+      assertEquals(group.getBinary("Name", 0).length(), 100);
+      assertEquals(group.getBinary("Gender", 0).length(), 100);
+      Group subGroup = group.getGroup("Links", 0);
+      assertEquals(subGroup.getBinary("Backward", 0).length(), 100);
+      assertEquals(subGroup.getBinary("Forward", 0).length(), 100);
```

Comment:
I don't like these tests. You should test for exact values. You may either keep the generated data in memory so you can match the file content with it or you may read both the source and the target file in the same time and match the values.

Reply:
I will cache the input data and compare it.